### ISSUE TITLE
How to prove to the user that my server is using the official files from the privatebin repository

### ISSUE THREAD

**Issue by ZerooCool** [PrivateBin/PrivateBin#647]
The subject has already been initiated somewhere, but, I have not consulted the 600 issues.

I'm trying to find out how to prove to the user that my server is using the official files from the privatebin repository.

I asked myself if it was possible to list the critical files, to compress them in a file, and, to generate a hash md5, or, other, which can be compared to a hash made available on the Github repository of privatebin, that , in order to always allow the user to make sure that no malicious file has come to corrupt the tool.

**Comment by elrido** [comment]
This is actually one of the use cases that the SRI hashes let you do. As you can see in the template files, all of the javascript files have a SRI hash generated and associated with them, here for example for the 1.3.4 release for the privatebin.js file that handles all of the encryption related logic:
https://github.com/PrivateBin/PrivateBin/blob/9914c37683496d161d9f8c8355bc89c358b1b243/tpl/bootstrap.php#L75

Any browser modern enough that it can run privatebin will refuse to load the application if you have modified the javascript file and not updated the SRI hash accordingly. A user of your server can therefore, theoretically, do the following:
1. Visit your privatebin instance in a browser
2. open the websites source code view, i.e. by pressing [Ctrl]+[U] or using the web developer console
3. Search for privatebin.js and compare the SRI hash with that of the release in question - for 1.3.4 it is `sha512-+4ay2Dw/9w/DOTutdpoiHvF6RxqV0V8ABSXHlebjGbQ7Td+3If7mOZ5OIu/lWCifcgP6vyhJuiB8+jl8FGtuwg==`
4. If it matches, they know you didn't manipulate the script and can assume that their browser keeps using the cached version of that file at least during that session

In practice, this is obviously not a very useful solution:
- You might not be using a release version, but something from the development branch here on github
- This is a very technical way of checking that, so it is unlikely they would do so every time. The hash and file might be different later on and they wouldn't notice.
- Even if that file is unchanged, you could just add another JS file with some malicious code and even we as developers of the software wouldn't necessarily notice that. All of the loaded JS files share of a website share a sandbox, so such an additional JS could for example be used to record and forward the secret key and password to you.
- There is a fundamental issue in terms of having to trust the server when it delivers you the encryption logic - this can't be avoided with javascript in the browser as far as we are aware. For such scenarios you do need to use native applications for encryption. See [point 4.1 in this report](https://privatebin.info/reports/zerobin-audit.html) for further details.

The bottom line is that users will have to trust their server, when using privatebin. We therefore encourage anyone to run their own instance, if possible.

**Comment by ZerooCool** [comment]
My JS file: https://privatebin.visionduweb.fr/js/privatebin.js
Vversion 1.3.4 de PrivateBin : sha512-+4ay2Dw/9w/DOTutdpoiHvF6RxqV0V8ABSXHlebjGbQ7Td+3If7mOZ5OIu/lWCifcgP6vyhJuiB8+jl8FGtuwg==

Use https://www.srihash.org ?
The URL does not support Cross-Origin Resource Sharing (CORS), when it should send a response header like Access-Control-Allow-Origin: *

Use https://report-uri.com/home/sri_hash ?
The server does not have CORS enabled. This resource can't be integrity checked.

Use https://zinoui.com/tools/sri-generator ?
sha512-pIM3/kBh5nI0TNPXyjhhqCrLE2enQc0DjiIfmpH2RoHaeDPNh2aTKIqEf8Ms6JMGWD/xJreAU7XUElWgELCkYQ==
SRI not good.

How i can check the SRI ?
I'm looking for.

**Comment by elrido** [comment]
AFAIK caniuse.com is correct: If a resource is loaded in a browser with SRI support and the resource include in the HTML has a SRI attached to it, the browser will download but refuse to execute/apply the JS/CSS in question, if there is a mismatch in the hash vs the file that is received. For Privatebin this means that you would see this message in your browser, if the privatebin.js can't be loaded for this reason:
https://github.com/PrivateBin/PrivateBin/blob/e35a26c9166d85b615a9c696c028a5cde95481c3/tpl/bootstrap.php#L572-L573

If you instead get to see a working instance, then:
1. You have a browser that is modern enough to support SRI, because we make use of much later features like web assembly and the web crypto API.
2. It would still be possible that the instance in question has removed the SRI hashes in their template, then the can't be enforced and they could send you a modified privatebin.js and it would work fine. Hence the suggestion to just open the source view and check that an SRI hash is provided and you can therefore rely on being enforced by your browser.

Finally, regarding the Content-Security-Header, this is indeed used by PrivateBin and I recommend not overriding it in your VirtualHost settings - the webserver settings take precedence over the application behind it. It is sent here:
https://github.com/PrivateBin/PrivateBin/blob/e35a26c9166d85b615a9c696c028a5cde95481c3/lib/Controller.php#L348
And you can change it if necessary in the configuration file:
https://github.com/PrivateBin/PrivateBin/blob/e35a26c9166d85b615a9c696c028a5cde95481c3/cfg/conf.sample.php#L82

**Comment by ZerooCool** [comment]
I understand without understanding.
The point that bothers me is that the verification is not transparent for the end user.
it works, or does not work, but, the user does not play any role for the verification.
I don't know much about SRI, but, is it possible to test for myself the sum of hash of a file, to see if it corresponds to the content of a file, without having to simply observe if the site works?

As said, I found some SRI generators but no validating SRI to compare the validity of a file for a given hash.

There is my Vhost : https://wiki.visionduweb.fr/index.php?title=VirtualHosts_des_domaines_enregistr%C3%A9s#privatebin.visionduweb.fr_.C3.A9coute_du_port_SSL_443

**Comment by rugk** [comment]
>  is it possible to test for myself the sum of hash of a file

Yes. There are many generators online or CLI tools etc.

----

There is nothing we can do about that to make it more transparent for the user or so. This is a browser limitation.

To be clear, as stated many times already: If a webserver admin wants to introduce malicious JS code on their PrivateBin instance, they can. Period.
SRI cannot and is not supposed to prevent this.

Also see #627 for the idea for a browser add-on to verify the SRI hashes externally, which is the only idea we had that _could_ theoretically work.

Yet again, see the Readme and FAQ for more information.

So as it seems this question has been answered and I thus close this issue. If anyone has further questions on that topic, feel free to comment here, again.

**Comment by ZerooCool** [comment]
SRI Validation srisum-rs project : https://github.com/zkat/srisum-rs/

**Comment by rugk** [comment]
@ZerooCool Yeah, as far as I see this just calculates the SRI hashes. This is not a problem and can also "easily" be done in JS.

**Comment by ZerooCool** [comment]
It should be possible to make available, from the official website, a form to enter a domain URL using Privatebin, which tests the integrity of the IRS, indicating whether the IRS is valid and up to date, or, if the IRS is from an old version of an old file, or, if the IRS is not valid.

The main objective would be to be able to allow a user to be reassured about the integrity of the privatebin that he would use?

**Comment by rugk** [comment]
This would be possible, but the security advantage is not very high.
Because if you put in enough effort, you could serve a different version to that "testing service" than to your users.
Generally, a one-time test cannot assure the next time you (or someone else) accesses the website that the code has not been tampered with.

As such, it is a much better idea to integrate that into a browser extension: https://github.com/PrivateBin/PrivateBin/issues/627

**Comment by ZerooCool** [comment]
Indeed, you're right, it's the same principle for robots.txt that we can serve differently, depending on the domain.
